src/openms/source/ANALYSIS/OPENSWATH/DIAPrescoring.cpp

```diff
@@ -130,59 +130,67 @@ namespace OpenMS
   void DiaPrescore::score(OpenSwath::SpectrumPtr spec,
                           const std::vector<OpenSwath::LightTransition>& lt,
                           double& dotprod,
-                          double& manhattan)
+                          double& manhattan) const
   {
     std::vector<std::pair<double, double> > res;
-    getMZIntensityFromTransition(lt, res);
-    std::vector<double> firstIstotope, theomasses;
-    DIAHelpers::extractFirst(res, firstIstotope);
-    std::vector<std::pair<double, double> > spectrum, spectrum2;
-    DIAHelpers::addIsotopes2Spec(res, spectrum, nr_charges_);
-    spectrum2.resize(spectrum.size());
-    std::copy(spectrum.begin(), spectrum.end(), spectrum2.begin());
-    //std::cout << spectrum.size() << std::endl;
-    DIAHelpers::addPreisotopeWeights(firstIstotope, spectrum, 2, 0.0);
-    //extracts masses from spectrum
-    DIAHelpers::extractFirst(spectrum, theomasses);
-    std::vector<double>  theorint;
-    DIAHelpers::extractSecond(spectrum, theorint);
+    std::vector<std::pair<double, double> > spectrumWIso, spectrumWIsoNegPreIso;
+    int chg = 1;
+    for (const auto& transition : lt)
+    {
+      if (transition.fragment_charge != 0) chg = transition.fragment_charge;
+      DIAHelpers::addIsotopes2Spec(transition.getProductMZ(), transition.getLibraryIntensity(), spectrumWIso, nr_isotopes_, chg);
+    }
+    spectrumWIsoNegPreIso.resize(spectrumWIso.size());
+    std::copy(spectrumWIso.begin(), spectrumWIso.end(), spectrumWIsoNegPreIso.begin());
+    double totalNegWeight = 1. / nr_isotopes_; // how much of ONE transition should be negatively weighted at the prePeaks (distributed equally on them)
+    UInt nrNegPeaks = 2;
+    for (const auto& transition : lt)
+    {
+      if (transition.fragment_charge != 0) chg = transition.fragment_charge;
+      DIAHelpers::addPreisotopeWeights(transition.getProductMZ(), spectrumWIso, nrNegPeaks, 0.0);
+      DIAHelpers::addPreisotopeWeights(transition.getProductMZ(),
+                                       spectrumWIsoNegPreIso,
+                                       nrNegPeaks,
+                                       -totalNegWeight / (double(lt.size() * nrNegPeaks)));
+    }
+    DIAHelpers::sortByFirst(spectrumWIso);
+    DIAHelpers::sortByFirst(spectrumWIsoNegPreIso);
```

Comment:
this seems like the main difference to the previous implementation
can you comment that part?